Improve perf scanning forward and backward through images

scottwittenburg opened the pull request:
By building up navigation links within the session tree once, when the sessions are loaded, we can avoid the triple-nested loops through the session tree each time we click a navigation button.  

This seems to result in a ballpark 3x performance improvement when advancing through images in a time series.

scottwittenburg commented:
@jimklo Is there any way you can try this out and see if you get the same performance improvement I experienced?

scottwittenburg commented:
Just added you as a reviewer @BryonLewis, at the request of @aashish24.

BryonLewis commented:
I know this is still on my radar, I've just been a bit busy with some other projects.

BryonLewis commented on client/src/store/index.js:
```diff
@@ -185,23 +118,59 @@ const store = new Vuex.Store({
+      let previousDataset = null;
+      let firstInPrev = null;
+
+      for (let i = 0; i < state.sessionTree.length; i++) {
+        let experiment = state.sessionTree[i];
+        for (let j = 0; j < experiment.sessions.length; j++) {
+          let session = experiment.sessions[j];
+          for (let k = 0; k < session.datasets.length; k++) {
+            let dataset = session.datasets[k];
+            dataset.session = session;
+            dataset.index = k;
+            dataset.previousDataset = previousDataset;
+            dataset.firstDatasetInPreviousSession = firstInPrev;
+            previousDataset = dataset;
+          }
+          firstInPrev = session.datasets[0];
+        }
```
This is just a general note that Vuex isn't really designed for complex nested objects.  The standard is try to normalize them into a flattened array.  This is another very minor thing that has to do with Vuex and how it works.

scottwittenburg replied:
Yeah, thanks, I've been wondering about that.  Mostly from the perspective of performance when moving through the timesteps in a scan (single arrow buttons).  For this PR, I'd be willing to spend the effort to re-think this bit of state if it could buy us some performance improvement.  

So if you think that flattening this data structure out has a chance of speeding up scanning through datasets, let me know, and I'll move it to the top of my current todo list (still a follow-on PR unless you object).

BryonLewis reviewed (commented):
I have mostly general comments.  I pulled and tested with the sample data and adding some additional repeats to just see how the advancing through data and sessions works.  Everything seems to work better right now and obviously the idea of building up a tree is better than doing the search request on each click.

Most of my comments have to do with what I believe is some of the legacy setup of the Vuex system and utilizing it a bit differently for better debugging.  If you use mutations for everything the state in the the Vue developer tools extensions would properly update and display the current information.  Right now it doesn't seem to do that because of the direct manipulation of state (which does work for reactivity on the page).

It's entirely up to you if you want to change stuff right now or just have a future PR to address some of these.  I recognize that having the updated JSON and the fix to the requirements are more important, and the changes to the session scanning improve the end user experience.  Most of my comments are just general vuex usage and not anything related to this specific PR.

scottwittenburg commented on client/src/store/index.js:
```diff
@@ -247,7 +216,7 @@ const store = new Vuex.Store({
         state.errorLoadingDataset = true;
       } finally {
-        state.currentDatasetId = dataset["_id"];
+        state.currentDataset = dataset;
```
Thank you @BryonLewis these are great tips.  I have noticed the missing mutations for the application state, but have not yet taken the time to make a pass to clean that up.  I would like to do it, but I agree another PR may be more appropriate.

scottwittenburg commented on server/setup.py:
```diff
@@ -6,7 +6,7 @@
     'girder==3.0.3',
     'girder_jobs==3.0.3',
-    'girder-autojoin==3.0.3'
+    'girder-autojoin==3.0.3',
```
Indeed, that's correct.

scottwittenburg commented:
Thanks for taking the time to look through this, test it out, and provide some thoughtful comments @BryonLewis, I really appreciate it.

As long as you have no objections, I'd like to merge this, and then start addressing the issues you brought in a subsequent PR.

BryonLewis reviewed (approved):
All good, I just wanted to make sure that my comments were saw before it squashed.
In terms of a flatter structure for the session data It all depends on how much of a headache you want to go through for the performance increase and at some point I would bet that image retrieval and loading is the bottle neck.

scottwittenburg commented:
Thanks @BryonLewis I'll merge this.

> at some point I would bet that image retrieval and loading is the bottle neck

That is a bottleneck indeed, but the performance we're looking for right now is not affected by that.  When stepping through images in a time series (using the single arrows or associated slider), all the images in the series are already fetched and loaded into memory in the application.  You can verify that by opening the network tab while you're scanning through time steps.  If you do not have access to a dataset with a long time series, I can get you one if you want.

Thanks again.

BryonLewis commented:
Thanks, that would help for the future.  I didn't have a large data set and only had the sample data.  I actually had an email typed up asking if you had a large dataset in mind for specific performance testing.  I edited the JSON and copied some files to create a slightly larger set of  items, but didn't really know if performance was relative on 10's, 100's or 1000's of items.

scottwittenburg commented:
> I actually had an email typed up asking if you had a large dataset in mind for specific performance testing

Ok, I got the ball rolling on this.
